helpers/embeds.js

```diff
@@ -0,0 +1,102 @@
+import { html, nothing, render } from 'lit';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { map } from 'lit/directives/map.js';
+import { tryGetPluginByKey } from './plugins.js';
+
+const embedTypeAttributeName = 'data-d2l-embed-type';
+
+function tryGetEmbedRendererPlugin(embedType) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) {
+		console.warn(`d2l-html-embed-renderer: Can't find plugin for ${embedType} embed type.`);
+	}
+
+	return embedRendererPlugin;
+}
+
+export async function createEmbedViewPlaceholder(embedType, props) {
+	const embedRendererPlugin = tryGetEmbedRendererPlugin(embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getViewPlaceholderData(props);
+	const contents = placeholderData.contents
+		? map(Object.entries(placeholderData.contents), ([id, content]) => html`<template data-d2l-embed-template-id=${id}>${content}</template>`)
+		: nothing;
+
+	const properties = placeholderData.properties ? JSON.stringify(placeholderData.properties) : undefined;
+
+	return placeholderData.inline
+		? html`
+			<span data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(properties)}" style="${placeholderData.style}">
+				${contents}
+			</span>
+		` : html`
+			<div data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(properties)}" style="${placeholderData.style}">
+				${contents}
+			</div>
+		`;
+}
+
+export async function createEmbedEditorPlaceholder(embedType, props) {
+	const embedRendererPlugin = tryGetEmbedRendererPlugin(embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getEditorPlaceholderData(props);
+	const text = placeholderData.text || nothing;
+
+	const properties = placeholderData.properties ? JSON.stringify(placeholderData.properties) : undefined;
+
+	return placeholderData.inline
+		? html`
+			<span
+				contentEditable="${ifDefined(placeholderData.contentEditable)}"
+				data-d2l-embed-type="${embedType}"
+				data-d2l-embed-props="${ifDefined(properties)}"
+				style="${placeholderData.style}">
+				${text}
```

Comment:
This supports the current @mentions case, but I didn't want to prototype anything further. It doesn't necessarily matter if we get this one right immediately, since the in-editor placeholder will never get saved anywhere.